Wikipedia/Code/WMFColumnarCollectionViewLayout.m

```diff
@@ -0,0 +1,165 @@
+@property (nonatomic) BOOL needsNewLayout;
+@property (nonatomic) BOOL needsLayout;
+
+@end
+
+@implementation WMFColumnarCollectionViewLayout
+
+- (nonnull instancetype)initWithMetrics:(nonnull WMFCVLMetrics *)metrics {
+    self = [super init];
+    if (self) {
+        self.metrics = metrics;
+        self.needsNewLayout = YES;
+        self.needsLayout = NO;
+    }
+    return self;
+}
+
+- (instancetype)init {
+    return [self initWithMetrics:[WMFCVLMetrics defaultMetrics]];
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self = [super initWithCoder:aDecoder];
+    if (self) {
+        if (!self.metrics) {
+            self.metrics = [WMFCVLMetrics defaultMetrics];
+        }
+    }
+    return self;
+}
+
+#pragma mark - Classes
+
++ (Class)invalidationContextClass {
+    return [WMFCVLInvalidationContext class];
+}
+
++ (Class)layoutAttributesClass {
+    return [WMFCVLAttributes class];
+}
+
+#pragma mark - Properties
+
+- (id <WMFColumnarCollectionViewLayoutDelegate>)delegate {
+    assert(self.collectionView.delegate == nil || [self.collectionView.delegate conformsToProtocol:@protocol(WMFColumnarCollectionViewLayoutDelegate)]);
+    return (id <WMFColumnarCollectionViewLayoutDelegate>)self.collectionView.delegate;
+}
+
+- (NSInteger)numberOfItemsInSection:(NSInteger)section {
+    return [self.collectionView.dataSource collectionView:self.collectionView numberOfItemsInSection:section];
+}
+
+- (CGSize)collectionViewContentSize {
+    return self.info.contentSize;
+}
+
+#pragma mark - Layout
+
+- (nullable NSArray<__kindof UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect {
+    
+    NSMutableArray *attributesArray = [NSMutableArray array];
+    
+    [self.info enumerateSectionsWithBlock:^(WMFCVLSection * _Nonnull section, NSUInteger idx, BOOL * _Nonnull stop) {
+        if (CGRectIntersectsRect(section.frame, rect)) {
+            [section enumerateLayoutAttributesWithBlock:^(WMFCVLAttributes *attributes, BOOL *stop) {
+                if (CGRectIntersectsRect(attributes.frame, rect)) {
+                    [attributesArray addObject:attributes];
+                }
+            }];
+        }
+    }];
+    
+    return attributesArray;
+}
+
+- (nullable UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
+    return [self.info layoutAttributesForItemAtIndexPath:indexPath];
+}
+
+- (nullable UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath {
+    return [self.info layoutAttributesForSupplementaryViewOfKind:elementKind atIndexPath:indexPath];
+}
+
+- (nullable UICollectionViewLayoutAttributes *)layoutAttributesForDecorationViewOfKind:(NSString*)elementKind atIndexPath:(NSIndexPath *)indexPath {
+    return nil;
+}
+
+- (void)prepareLayout {
+    if (self.needsLayout || self.needsNewLayout) {
+        if (self.needsNewLayout) {
```

Comment:
I'm moving the needsNewLayout flag and other changes into a follow on PR since they're going beyond parity with the table view - but yeah, the info should stay valid until this point. I'll try some other things in the follow on.